HW2/P2/parallel_vector.pyx

```diff
@@ -105,11 +112,17 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
    # to parallelize data movement.  Be sure to avoid deadlock, as well as
    # double-locking.
    ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+   for r in range(repeat):
+       for idx in range(src.shape[0]):
+           if src[idx] == dest[idx]:
+               continue
+           if counts[src[idx]] > 0:
+               acquire(&(locks[dest[idx]/N]))
+               counts[dest[idx]] += 1
+               release(&(locks[dest[idx]/N]))
+
+               acquire(&(locks[src[idx]/N]))
```

Comment:
I would suggest dividing the number of locks before once, instead of at every call. This gets rid of multiple operation in the loop and could substantially increase your computation speed.